F OpenNebula/one#6408: Update docs (#320)
OpenNebula/website: content/product/operation_references/hypervisor_configuration/kvm_driver.md

- `DISK`: attributes `DRIVER`, `CACHE`, `IO`, `DISCARD`, `TOTAL_BYTES_SEC`, `TOTAL_BYTES_SEC_MAX`, `TOTAL_BYTES_SEC_MAX_LENGTH`, `TOTAL_IOPS_SEC`, `TOTAL_IOPS_SEC_MAX`, `TOTAL_IOPS_SEC_MAX_LENGTH`, `READ_BYTES_SEC`, `READ_BYTES_SEC_MAX`, `READ_BYTES_SEC_MAX_LENGTH`, `WRITE_BYTES_SEC`, `WRITE_BYTES_SEC_MAX`, `WRITE_BYTES_SEC_MAX_LENGTH`, `READ_IOPS_SEC`, `READ_IOPS_SEC_MAX`, `READ_IOPS_SEC_MAX_LENGTH`, `WRITE_IOPS_SEC`, `WRITE_IOPS_SEC_MAX`, `WRITE_IOPS_SEC_MAX_LENGTH`, `SIZE_IOPS_SEC`.
- `NIC`: attribute `FILTER`, `MODEL`.
- `GRAPHICS`: attributes `TYPE`, `LISTEN`, `PASSWD`, `KEYMAP`, `RANDOM_PASSWD`. The VM instance must have at least empty `GRAPHICS = []` section to read these default attributes from the config file and to generate cluster unique `PORT` attribute.
- `VIDEO`: attributes: `TYPE`, `IOMMU`, `ATS`, `VRAM`, `RESOLUTION`.
- `RAW`: to add libvirt attributes to the domain XML file.
- `HYPERV_OPTIONS`: to enable hyperv extensions.
- `HYPERV_TIMERS`: timers added when HYPERV is set to yes in FEATURES.
- `SPICE_OPTIONS`: to add default devices for SPICE.

The following attributes can be overridden at Cluster and Host level, but not within individual VM configuration:

> - `OVMF_UEFIS`: to add allowed file paths for Open Virtual Machine Firmware

Reviewer: ```suggestion
- `OVMF_UEFIS`: to specify allowed file paths for Open Virtual Machine Firmware
```

- `Q35_ROOT_PORTS`: to modify the number of PCI devices that can be attached in q35 VMs (defaults to 16)

> - `Q35_NUMA_PCIE`: Generate a NUMA-aware PCIe topology for pinned VMs

Reviewer: ```suggestion
- `Q35_NUMA_PCIE`: to generate a NUMA-aware PCIe topology for pinned VMs
```

> - `CGROUPS_VERSION`: Use ‘2’ to use Cgroup V2, all other values or undefined: use Cgroup V1

Reviewer: What's the purpose of this attribute? The description indicates values, but no goals as other attributes.

Reviewer 2: This is used to select the version, VERSION = 2 means use that version. I think in this context should be self-explanatory

> - `EMULATOR_CPUS`: Value used for KVM option <cputune><emulatorpin cpuset=…>

Reviewer 2: Agree we should include a brief note on what about something like:

specifies which of host physical CPUs will run the main QEMU emulator thread

Optionally, you can set up cgroups to control resources on your Hosts. By default KVM VMs will be placed in the `machine.slice`, the resources assigned in this slice can be adjusted for each hypervisor. The [libvirt cgroups documentation](https://libvirt.org/cgroups.html) describes all the cases and the way the cgroups are managed by libvirt/KVM.

> OpenNebula will compute the `shares` attribute of the Libvirt domain using the `CPU` parameter and the base share value, which depends on the cgroups version of the hypervisor. For example, a VM with `CPU=2` will get a cgroup value of `cpu.shares = 2048` (or `cpu.weight=200` for cgroups version 2), twice the default value. Note that if you have a mix of cgroups version 1 and 2 Hosts, you may have inconsistent resource distribution when live-migrating a VM across different versions.

Reviewer 2: a. it is the base shared value
b. either, depending on the cgroups version in the hypervisor

> OpenNebula will compute the `shares` attribute of the Libvirt domain using the `CPU` parameter and the base share value, which depends on the cgroups version of the hypervisor. For example, a VM with `CPU=2` will get a cgroup value of `cpu.shares = 2048` (or `cpu.weight=200` for cgroups version 2), twice the default value. Note that if you have a mix of cgroups version 1 and 2 Hosts, you may have inconsistent resource distribution when live-migrating a VM across different versions.

Reviewer: Thanks! Let me edit the suggestion.

#### DISK

> - `TYPE`: This attribute defines the type of media to be exposed to the VM; possible values are `disk` (default) or `cdrom`. This attribute corresponds to the `media` option of the `-driver` argument of the `kvm` command.

Reviewer: ```suggestion
- `TYPE`: this attribute defines the type of media to be exposed to the VM; possible values are `disk` (default) or `cdrom`. This attribute corresponds to the `media` option of the `-driver` argument of the `kvm` command.
```

### Enabling QEMU Guest Agent

QEMU Guest Agent allows the communication of some actions with the guest OS. This agent uses a virtio serial connection to send and receive commands. One of the interesting actions is that it allows you to freeze the filesystem before doing a snapshot. This way the snapshot won’t contain half-written data. Filesystem freeze will only be used with `CEPH` and `qcow2` storage drivers.

The agent package needed in the Guest OS is available in most distributions. It’s called `qemu-guest-agent` in most of them. If you need more information you can follow these links:

- [QEMU Guest Agent - libvirt](http://wiki.libvirt.org/page/Qemu_guest_agent)

> - [QEMU Guest Agent - rhel](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Virtualization_Deployment_and_Administration_Guide/chap-QEMU_Guest_Agent.html)

Reviewer: ```suggestion
- [QEMU Guest Agent - rhel](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_and_managing_virtualization/assembly_enabling-virtualization-in-rhel-9_configuring-and-managing-virtualization#enabling-qemu-guest-agent-features-on-your-virtual-machines_assembly_enabling-virtualization-in-rhel-9)
```

The driver consists of the following files:

- `/usr/lib/one/mads/one_vmm_exec` : generic VMM driver.
- `/var/lib/one/remotes/vmm/kvm` : commands executed to perform actions.

And the following driver configuration files:

> - `/etc/one/vmm_exec/vmm_exec_kvm.conf` : This file contains default values for KVM domain definitions (in other words, OpenNebula templates). It is generally a good idea to configure defaults for the KVM-specific attributes here, that is, attributes mandatory in the KVM driver that are not mandatory for other hypervisors. Non-mandatory attributes for KVM but specific to them are also recommended to have a default. Changes to this file **require opennebula to be restarted**.

Reviewer: ```suggestion
- `/etc/one/vmm_exec/vmm_exec_kvm.conf` : this file contains default values for KVM domain definitions;  in other words, OpenNebula templates. It is a good practice to configure defaults for the KVM-specific attributes here, that is, attributes mandatory in the KVM driver that are not mandatory for other hypervisors. Non-mandatory attributes for KVM but specific to them are also recommended to have a default. Changes to this file **require opennebula to be restarted**.
```

> * `/var/lib/one/remotes/etc/vmm/kvm/kvmrc` : This file holds instructions to be executed before the actual driver load to perform specific tasks or to pass environmental variables to the driver. The syntax used for the former is plain shell script that will be evaluated before the driver execution. For the latter, the syntax is the familiar:

Reviewer: ```suggestion
* `/var/lib/one/remotes/etc/vmm/kvm/kvmrc` : this file holds instructions to be executed before the actual driver load to perform specific tasks or to pass environmental variables to the driver. The syntax used for the former is plain shell script that will be evaluated before the driver execution. For the latter, the syntax is the familiar:
```
